ChordRatioTool

david-hall opened the pull request:
Play with it a little and let me know what you think.

vorth commented:
Awesome work!

As you probably guessed, I don't think we can use up the toolbar space with five new built-in tools, though I appreciate the work on adding text overlays!  Instead, I think we need to replace the "Construct/Affine Pentagon" menu item with a separator and five new items, filtered per field support of course.

(Someday, we'll have a better toolbar management scheme, and we can have more built-in tools for people to add to their custom toolbar.  Though even then, I think the five you defined should have more specific and intuitive icons.)

Since the tool is clearly working out where to add a new strut, I think it should leave the new strut and its antecedent neighbor selected.  Furthermore, it should require a selection of exactly two coincident struts, and it should throw an error for any other selection.

I notice that the "show and select parameters" button isn't working.

david-hall commented:
Thanks for the feedback. I knew that toolbar real estate was going to be a
problem but I decided to try it anyway, mainly to get more familiar with
how to integrate new tools. I like your idea of accessing the built-in
tools from a sub-menu instead of the tool bar. The fact that I added
CHORD_RATIO as a new Tool.Kind will probably help to simplify that approach.

I had not noticed that the "show and select parameters" button isn't
working. I have fixed that in a new commit.

I wanted to require the inputs to be exactly two struts with a common
endpoint as you suggest but I was having trouble controlling which balls
and struts get selected and unselected. That appears to be an
all-or-nothing setting in the tool configuration and it gets applied
outside of the tool itself. Now that the basic tool functionality is
stable, maybe I should revisit that aspect. Also, I intentionally ignore
any pre-selected balls and leave them selected so that I can apply the tool
repeatedly and leave a trail of balls which can be inputs to JoinBalls or a
panel or a 2D convex hull. The input struts remaining selected are not a
problem for any of those tools although I agree that it would be nice if I
could deselect all but the two that are needed for the next iteration. The
built-in tools have a pre determined repeat count which allows an entire
affine polygon to be generated with one click. The user-defined tools and
the built-in parabola tool have a non-configurable repeat count of 1.
That's why I leave all of the balls selected. Similarly, I can ignore
selected panels as they have no effect.

On a related note, check out my new StrutDivisionTool. It's similar to the
ChordRatioTool in that it takes up a lot of toolbar real estate. It would
also benefit from being accessed via the menu. I won't submit it as a PR
yet, but if you get a chance, check it out.  It's on its own new git
branch. One aspect of the StrutDivision that I like is that it not only
divides a strut in two new struts of a given proportion, but it can also
divide a strut into more than two parts by selecting up to 10 struts from
the same orbit as inputs. The built-in tools include some examples of a
3-way split. Any field of order 3 or more has a built-in tool that splits
proportionally for each term.  For example, a heptagon field generates
tools that will divide a strut into 1:rho:sigma and since the division
respects the orientation of the strut, there's also another tool that will
divide a strut into the reverse order of sigma:rho:1. Although I haven't
tried this tool with my new polygon fields, it's all designed to let any of
them generate similar splits automatically which would really fill up the
toolbar.

Another thing you may notice is that the new struts and balls generated
with the ChordRatioTool use system colors rather than copying the color of
either input strut (besides the fact that the two input struts may be
different orbits or different colors). That was mainly because I didn't
want the newly generated balls to take on the color of the input
struts, but also because I personally liked the new struts to use the
system colors. That's also another argument for allowing multiple struts as
inputs on the ChordRatioTool and leaving both the input and output struts
selected so you can apply the same color to all of them if you like. In the
case of the StrutDivisionTool, I let the tool copy the original color to
the output struts since they are in the same orbit and replace the input
strut, but the new balls are white rather than the color of the strut.
that's hard coded in the tool. I know you are working on making the
color-copy optional for tools, but these two tools highlight some use
cases where a set of input struts generates both balls and struts and the
copy color rules may not be a simple yes or no. Similarly, the
ProjectionTool can project a panel to a strut or a strut to a ball, but in
that case, it would make sense to copy the color since there is a 1:1 match
of inputs to outputs. We can discuss some of these color copy use cases
whenever you like. I just wanted to mention these ideas while they're on my
mind.

I'm going to experiment with getting these predefined tools off of the tool
bar and onto the menu. I'll keep you posted.



On Wed, Jun 23, 2021 at 7:04 AM Scott Vorthmann ***@***.***>
wrote:

> Awesome work!
>
> As you probably guessed, I don't think we can use up the toolbar space
> with five new built-in tools, though I appreciate the work on adding text
> overlays! Instead, I think we need to replace the "Construct/Affine
> Pentagon" menu item with a separator and five new items, filtered per field
> support of course.
>
> (Someday, we'll have a better toolbar management scheme, and we can have
> more built-in tools for people to add to their custom toolbar. Though even
> then, I think the five you defined should have more specific and intuitive
> icons.)
>
> Since the tool is clearly working out where to add a new strut, I think it
> should leave the new strut and its antecedent neighbor selected.
> Furthermore, it should require a selection of exactly two coincident
> struts, and it should throw an error for any other selection.
>
> I notice that the "show and select parameters" button isn't working.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/vZome/vzome/pull/393#issuecomment-866742524>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACTTECQVBJKRHT6F2YBEMBLTUG5T3ANCNFSM47CJHVVQ>
> .
>

david-hall commented:
Per our conversations, the entire approach for providing this functionality as built-in tools has been replaced by more appropriate dynamic menu selections using edits instead of tools. PR #415 includes that functionality so I am closing this PR as superfluous. At some point, I may revisit the use of a ChordRatioTool factory where the user can make custom chord ratios, for example, those that generate hyperbolic curves or have negative chord ratio values in addition to the dynamic menu options, but it will not include the built-in tools like this version of the tool factory does.
